cli/command_snapshot_copy_move_history.go

```diff
@@ -0,0 +1,232 @@
+
+		if dstSource == manifest.Source {
+			log(ctx).Debugf("%v is the same as destination, ignoring", dstSource)
+			continue
+		}
+
+		if snapshotExists(dstSnapshots, dstSource, manifest.StartTime) {
+			if isMoveCommand && !snapshotCopyOrMoveDryRun {
+				log(ctx).Infof("%v (%v) already exists - deleting source", dstSource, formatTimestamp(manifest.StartTime))
+
+				if err := rep.DeleteManifest(ctx, manifest.ID); err != nil {
+					return errors.Wrap(err, "unable to delete source manifest")
+				}
+			} else {
+				log(ctx).Infof("%v (%v) already exists", dstSource, formatTimestamp(manifest.StartTime))
+			}
+
+			continue
+		}
+
+		srcID := manifest.ID
+
+		log(ctx).Infof("%v %v (%v) => %v", getCopySnapshotAction(isMoveCommand), manifest.Source, formatTimestamp(manifest.StartTime), dstSource)
+
+		if snapshotCopyOrMoveDryRun {
+			continue
+		}
+
+		manifest.ID = ""
+		manifest.Source = dstSource
+
+		if _, err := snapshot.SaveSnapshot(ctx, rep, manifest); err != nil {
+			return errors.Wrap(err, "unable to save snapshot")
+		}
+
+		if isMoveCommand {
+			if err := rep.DeleteManifest(ctx, srcID); err != nil {
+				return errors.Wrap(err, "unable to delete source manifest")
+			}
+		}
+	}
+
+	return nil
+}
+
+func getCopySnapshotAction(isMoveCommand bool) string {
+	action := "copying"
+	if isMoveCommand {
+		action = "moving"
+	}
+
+	if snapshotCopyOrMoveDryRun {
+		action += " (dry run)"
+	}
+
+	return action
+}
+
+func getCopySourceAndDestination(rep repo.Repository) (si, di snapshot.SourceInfo, err error) {
+	si, err = snapshot.ParseSourceInfo(snapshotCopyOrMoveSource, rep.ClientOptions().Hostname, rep.ClientOptions().Username)
+	if err != nil {
+		return si, di, errors.Wrap(err, "invalid source")
+	}
+
+	if snapshotCopyOrMoveDestination == "" {
+		// no destination - assume current user@hostname
+		di.UserName = rep.ClientOptions().Username
+		di.Host = rep.ClientOptions().Hostname
+	} else {
+		di, err = snapshot.ParseSourceInfo(snapshotCopyOrMoveDestination, rep.ClientOptions().Hostname, rep.ClientOptions().Username)
+		if err != nil {
+			return si, di, errors.Wrap(err, "invalid destination")
+		}
+	}
+
+	if di.Path != "" && si.Path == "" {
+		// it is illegal to specify source without path, but destination with a path
+		// as it would result in multiple individual paths being squished together.
+		return si, di, errors.Errorf("path specified on destination but not source")
+	}
+
+	if di.UserName != "" && si.UserName == "" {
+		// it is illegal to specify source without username, but destination with a username
+		// as it would result in multiple individual paths being squished together.
+		return si, di, errors.Errorf("username specified on destination but not source")
+	}
+
+	return si, di, nil
+}
+
+func snapshotExists(snaps []*snapshot.Manifest, src snapshot.SourceInfo, startTime time.Time) bool {
+	for _, s := range snaps {
+		if !startTime.Equal(s.StartTime) {
+			continue
+		}
+
+		if src != s.Source {
+			continue
+		}
+
```

Comment:
Should this check other fields of the snapshot manifest, in particular that the `RootEntry` fields match?

It'd be good to avoid deleting the source manifest when moving a snapshot for which there is a target manifest with same `startTime` but pointing to a different root. Deleting the source manifest in this case may result in loosing the only reference to the source's root.

It'd be good to simply say the snapshot does not exist when the root entry does not match that of the source.
Not sure whether having two snapshots at the destination with the same `startTime` may cause issues when computing the retentions.

Reply:
done